Add request_render context method

PoignardAzur opened the pull request:
Remove some redundant request_paint calls

Fixes #585.

waywardmonkeys reviewed (commented):
This is difficult to review without having more knowledge about what each interaction is expected to do / permitted to do, so that you know which things need to be updated.

It sort of sucks, but it would almost be useful to have a one line comment at each spot saying what's getting updated or not (to give an indication of why).

PoignardAzur commented:
Yeah. Ideally we'd want some way to test that the request flags are set (eg that we don't have a change that affected layout but didn't set the request_layout flag), but it would probably be pretty painful to implement.

DJMcNab commented:
Should this be marked as fixing #585?

PoignardAzur commented:
Oh, right, knew I forgot something.

waywardmonkeys commented:
> Yeah. Ideally we'd want some way to test that the request flags are set (eg that we don't have a change that affected layout but didn't set the request_layout flag), but it would probably be pretty painful to implement.

What I mean is that in each spot, you've made a decision as to which updates are appropriate and there were some criteria that you used in making that decision. It might be good to note that at the various callsites in a comment so that future-us will know why each decision was made and what the intent was.

DJMcNab commented on masonry/src/widget/spinner.rs:
```diff
@@ -93,15 +93,13 @@ impl Widget for Spinner {
                     self.t = 0.0;
                 }
                 ctx.request_anim_frame();
-                ctx.request_paint();
```
This strikes me as strange - does `AnimFrame` implicitly request a repaint? It seems to me that there are reasonable animations where you wouldn't want that to be the case.

PoignardAzur replied:
Nope, you're right, the new test for the animated spinner was failing because of that.

(Hurray for screenshot tests!)

DJMcNab commented on masonry/examples/calc_masonry.rs:
```diff
@@ -176,7 +176,7 @@ impl Widget for CalcButton {
                 accesskit::Action::Default => {
                     ctx.submit_action(Action::Other(Box::new(self.action)));
-                    ctx.request_paint();
+                    ctx.request_paint_only();
```
I'm surprised this line is necessary at all. Nothing has changed which would require a repaint to happen.

PoignardAzur replied:
Right. Now that you mention it, the paint method of that widget is empty, so `request_paint` should never be called anyway.

DJMcNab commented on masonry/src/contexts.rs:
```diff
@@ -462,12 +462,39 @@ impl<'w> QueryCtx<'w> {
+
+    /// Request a [`paint`](crate::Widget::paint) pass.
+    ///
+    /// This doesn't request an accessibility pass, which usually isn't what you want.
```
? Yes it does?

PoignardAzur replied:
That's what I get for copy-pasting.

DJMcNab commented on masonry/src/widget/button.rs:
```diff
@@ -104,15 +104,15 @@ impl Widget for Button {
                 accesskit::Action::Default => {
                     ctx.submit_action(Action::ButtonPressed(PointerButton::Primary));
-                    ctx.request_paint();
+                    ctx.request_paint_only();
```
Again, here. How would you expect this paint to be different from the previous one?

DJMcNab commented on masonry/src/widget/checkbox.rs:
```diff
@@ -94,16 +92,15 @@ impl Widget for Checkbox {
 
     fn on_status_change(&mut self, ctx: &mut LifeCycleCtx, _event: &StatusChange) {
-        ctx.request_paint();
+        ctx.request_render();
```
Hmm, should this be the default implementation of `on_status_changed`?

PoignardAzur replied:
Maybe. I'd rather not have a default implementation at all for now.

PoignardAzur commented:
@waywardmonkeys I've added some comments to some of the call sites.

I don't want to comment *every* call site; not just because that would be redundant, but because I think redundant comments actually make code *less* understandable (they make the reader wonder if they missed something).

So for something like this:

```rust
    pub fn set_line_break_mode(&mut self, line_break_mode: LineBreaking) {
        self.widget.line_break_mode = line_break_mode;
        self.ctx.request_layout();
    }
```

I think it's self-explanatory why changing the line breaking mode of a label impacts its layout.

Some other notes:

- Some of the changes in the PR were `request_paint` -> `request_render` where the widget didn't actually benefit from changing the accessibility node, just because I did a project-wide rename. I think it's fine, request_render is a fine default, and it's okay if it schedules some redundant work. Still, I've reverted some of these changes to `request_paint_only`.
- The Portal/ScrollBar code is kind of cursed and I tried to touch the request flags as little as possible there and keep behavior unchanged. In the future we might change requests to be a little more narrow (eg compose instead of paint).

jaredoconnell commented on masonry/src/widget/portal.rs:
```diff
@@ -20,6 +20,7 @@ use crate::{
 
 // TODO - refactor - see https://github.com/linebender/xilem/issues/366
 // TODO - rename "Portal" to "ScrollPortal"?
+// TODO - Document which cases need request_layout, request_compose and request_render
```
Is there a reason this isn't being done in this PR?

PoignardAzur replied:
Because it's more work than the rest of the PR combined.

jaredoconnell reviewed (approved):
This is definitely an improvement. Should help reduce stale accessibility nodes.
I tested general (non-screen-reader) use, and I don't see any regressions.

waywardmonkeys commented on masonry/src/contexts.rs:
```diff
@@ -483,12 +483,36 @@ impl<'w> QueryCtx<'w> {
-    pub fn request_paint(&mut self) {
+    ///
+    /// This doesn't request an [`accessibility`](crate::Widget::accessibility) pass,
+    /// which usually isn't what you want.
```
I find this wording confusing with the multiple negatives. Same for the function below.

waywardmonkeys commented on masonry/src/widget/button.rs:
```diff
@@ -82,7 +82,8 @@ impl Widget for Button {
                 if !ctx.is_disabled() {
                     ctx.capture_pointer();
-                    ctx.request_paint();
+                    // Changes in pointer capture impact appearance, but not accessibility node
```
This is great and exactly what I was hoping for when I asked about adding some comments. Thanks!

DJMcNab commented on masonry/src/widget/button.rs:
```diff
@@ -104,15 +106,15 @@ impl Widget for Button {
     fn on_status_change(&mut self, ctx: &mut LifeCycleCtx, _event: &StatusChange) {
-        ctx.request_paint();
+        // Changes in hovered/focused status impact appearance, but not accessibility node
+        ctx.request_paint_only();
```
Changing the focus status *does* change the accessibility node - https://github.com/linebender/xilem/blob/3ba0df6064a48a92a3efab51455efd378776d33e/masonry/src/passes/accessibility.rs#L119.

However, requesting the accessibility update in that case is not the responsibility of this code, so this is still true.
One thing I'm not sure of - does this update actually happen?

PoignardAzur replied:
Good catch. It would have to happen in `run_update_focus_pass`, and it currently doesn't.

DJMcNab commented on masonry/src/widget/split.rs:
```diff
@@ -355,15 +355,18 @@ impl WidgetMut<'_, Split> {
         self.widget.draggable = draggable;
-        self.ctx.request_paint();
+        // Bar mutability impacts appearance, but not accessibility node
+        // TODO - This might change in a future implementation
```
I agree that this is suspicious
